passes/cmds/partition.cc

```diff
@@ -0,0 +1,501 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2020  Alberto Gonzalez <boqwxp@airmail.cc>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/yosys.h"
+#include "kernel/cost.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct PartitionPass : public Pass {
+	//Constructs a hypergraph from a flat techmapped module, partitions the hypergraph,
```

Comment:
I'm not sure I understand the "techmapped" restriction. Do you actually mean that the module is restricted to working on netlists with Yosys fine-grained cells like `$_AND_`?

Reply:
There are two restrictions:
1. Hierarchy is not supported, though it shouldn't be too hard to add.
2. The module must only contain cells that are contained in `cell_type_costs`.  See `kernel/costs.h` in this PR for the definition of `CellCosts::unit_gate_cost()` which has a list of cell types.  It is my understanding that the module generally must have been run through `techmap` for this to be the case.  TL;DR yes, I mean it's restricted to `$_AND_` etc.

Reply:
Right. I would say that "techmapped" is confusing here because most Yosys techmapping does *not* produce those gates--it is done as a part of synthesis flows that produce vendor cells. Something like "fine grained" or "gate level" would be easier to understand.

Or just not make that restriction a part of the design short description, since it's not fundamental--I would very much like `partition` to work for CXXRTL some day. So, mentioning that currently it only works on gate level designs in the long description works too IMO.

Reply:
Fair enough. And I agree, it would be great to extend it to be more general.